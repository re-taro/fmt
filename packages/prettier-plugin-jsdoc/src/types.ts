import type { ParserOptions } from "prettier";

export interface JsdocOptions {
  jsdocSpaces: number;
  jsdocPrintWidth?: number;
  jsdocDescriptionWithDot: boolean;
  jsdocDescriptionTag: boolean;
  jsdocVerticalAlignment: boolean;
  jsdocKeepUnParseAbleExampleIndent: boolean;
  /** @default "singleLine" */
  jsdocCommentLineStrategy: "singleLine" | "multiline" | "keep";
  jsdocSeparateReturnsFromParam: boolean;
  jsdocSeparateTagGroups: boolean;
  jsdocAddDefaultToDescription: boolean;
  jsdocCapitalizeDescription: boolean;
  jsdocPreferCodeFences: boolean;
  tsdoc: boolean;
  jsdocLineWrappingStyle: "greedy";
}

export interface AllOptions extends ParserOptions, JsdocOptions {}

interface LocationDetails {
  line: number;
  column: number;
}
interface Location {
  start: LocationDetails;
  end: LocationDetails;
}

export interface PrettierComment {
  type: "CommentBlock" | "Block";
  value: string;
  start: number;
  end: number;
  loc: Location;
}

export interface Token {
  type:
    | "CommentBlock"
    | "Block"
    | {
        label: string;
        keyword?: string;
        beforeExpr: boolean;
        startsExpr: boolean;
        rightAssociative: boolean;
        isLoop: boolean;
        isAssign: boolean;
        prefix: boolean;
        postfix: boolean;
        binop: null;
      };
  value: string;
  start: number;
  end: number;
  loc: Location;
}

export interface AST {
  start: number;
  end: number;
  loc: Location;
  errors: [];
  program: {
    type: "Program";
    start: number;
    end: number;
    loc: [];
    sourceType: "module";
    interpreter: null;
    body: [];
    directives: [];
  };
  comments: PrettierComment[];
  tokens: Token[];
}
