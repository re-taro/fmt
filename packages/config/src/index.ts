export default {
  extends: [
    "plugin:@re-taro/json",
    "plugin:@re-taro/markdown",
    "plugin:@re-taro/react",
    "plugin:@re-taro/svelte",
    "plugin:@re-taro/typescript",
    "plugin:@re-taro/vue",
    "plugin:@re-taro/yaml",
    "prettier",
  ],
};
