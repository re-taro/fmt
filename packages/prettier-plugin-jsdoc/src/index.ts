import type { Parser, ParserOptions, SupportOption } from "prettier";
import parserBabel from "prettier/plugins/babel";
import parserFlow from "prettier/plugins/flow";
import parserTypescript from "prettier/plugins/typescript";

import { getParser } from "./parser";
import type { JsdocOptions } from "./types";
import { findPluginByParser } from "./utils";

export const options = {
  jsdocSpaces: {
    name: "jsdocSpaces",
    type: "int",
    category: "jsdoc",
    default: 1,
    description: "How many spaces will be used to separate tag elements.",
  },
  jsdocDescriptionWithDot: {
    name: "jsdocDescriptionWithDot",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should dot be inserted at the end of description",
  },
  jsdocDescriptionTag: {
    name: "jsdocDescriptionTag",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should description tag be used",
  },
  jsdocVerticalAlignment: {
    name: "jsdocVerticalAlignment",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should tags, types, names and description be aligned",
  },
  jsdocKeepUnParseAbleExampleIndent: {
    name: "jsdocKeepUnParseAbleExampleIndent",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description:
      "Should unParseAble example (pseudo code or no js code) keep its indentation",
  },
  jsdocCommentLineStrategy: {
    name: "jsdocCommentLineStrategy",
    type: "choice",
    choices: [
      {
        since: "1.1.0",
        value: "singleLine",
        description: "Should compact single line comment, if possible",
      },
      {
        since: "1.1.0",
        value: "multiline",
        description: "Should compact multi line comment",
      },
      {
        since: "1.1.0",
        value: "keep",
        description: "Should keep original line comment",
      },
    ],
    category: "jsdoc",
    default: "singleLine",
    description: "How comments line should be",
  },
  jsdocSeparateReturnsFromParam: {
    name: "jsdocSeparateReturnsFromParam",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Add an space between last @param and @returns",
  },
  jsdocSeparateTagGroups: {
    name: "jsdocSeparateTagGroups",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Add an space between tag groups",
  },
  jsdocCapitalizeDescription: {
    name: "jsdocCapitalizeDescription",
    type: "boolean",
    category: "jsdoc",
    default: true,
    description: "Should capitalize first letter of description",
  },
  tsdoc: {
    name: "tsdoc",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should format as tsdoc",
  },
  jsdocPrintWidth: {
    name: "jsdocPrintWidth",
    type: "int",
    category: "jsdoc",
    default: undefined as any,
    description:
      "If You don't set value to jsdocPrintWidth, the printWidth will be use as jsdocPrintWidth.",
  },
  jsdocAddDefaultToDescription: {
    name: "jsdocAddDefaultToDescription",
    type: "boolean",
    category: "jsdoc",
    default: true,
    description: "Add Default value of a param to end description",
  },
  jsdocPreferCodeFences: {
    name: "jsdocPreferCodeFences",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: `Prefer to render code blocks using "fences" (triple backticks). If not set, blocks without a language tag will be rendered with a four space indentation.`,
  },
  jsdocLineWrappingStyle: {
    name: "jsdocLineWrappingStyle",
    type: "choice",
    choices: [
      {
        since: "0.3.39",
        value: "greedy",
        description: "Lines wrap as soon as they reach the print width",
      },
    ],
    category: "jsdoc",
    default: "greedy",
    description:
      "Strategy for wrapping lines for the given print width. More options may be added in the future.",
  },
} satisfies Record<keyof JsdocOptions, SupportOption>;

export const defaultOptions: JsdocOptions = {
  jsdocSpaces: options.jsdocSpaces.default,
  jsdocPrintWidth: options.jsdocPrintWidth.default as unknown as undefined,
  jsdocDescriptionWithDot: options.jsdocDescriptionWithDot.default as boolean,
  jsdocDescriptionTag: options.jsdocDescriptionTag.default as boolean,
  jsdocVerticalAlignment: options.jsdocVerticalAlignment.default as boolean,
  jsdocKeepUnParseAbleExampleIndent: options.jsdocKeepUnParseAbleExampleIndent
    .default as boolean,
  jsdocCommentLineStrategy: options.jsdocCommentLineStrategy
    .default as "singleLine",
  jsdocSeparateReturnsFromParam: options.jsdocSeparateReturnsFromParam
    .default as boolean,
  jsdocSeparateTagGroups: options.jsdocSeparateTagGroups.default as boolean,
  jsdocCapitalizeDescription: options.jsdocCapitalizeDescription
    .default as boolean,
  jsdocAddDefaultToDescription: options.jsdocAddDefaultToDescription
    .default as boolean,
  jsdocPreferCodeFences: options.jsdocPreferCodeFences.default as boolean,
  tsdoc: options.tsdoc.default as boolean,
  jsdocLineWrappingStyle: options.jsdocLineWrappingStyle.default as "greedy",
};

export const parsers = {
  // JS - Babel
  get babel() {
    const parser = parserBabel.parsers.babel;

    return mergeParsers(parser, "babel");
  },
  get "babel-flow"() {
    const parser = parserBabel.parsers["babel-flow"];

    return mergeParsers(parser, "babel-flow");
  },
  get "babel-ts"() {
    const parser = parserBabel.parsers["babel-ts"];

    return mergeParsers(parser, "babel-ts");
  },
  // JS - Flow
  get flow() {
    const parser = parserFlow.parsers.flow;

    return mergeParsers(parser, "flow");
  },
  // JS - TypeScript
  get typescript(): Parser {
    const parser = parserTypescript.parsers.typescript;

    return mergeParsers(parser, "typescript");
    // require("./parser-typescript").parsers.typescript;
  },
  get "jsdoc-parser"() {
    // Backward compatible, don't use this in new version since 1.0.0
    const parser = parserBabel.parsers["babel-ts"];

    return mergeParsers(parser, "babel-ts");
  },
};

function mergeParsers(originalParser: Parser, parserName: string) {
  const jsDocParse = getParser(originalParser.parse, parserName) as any;

  const parser = {
    ...originalParser,
    preprocess: jsDocPreprocess,
    parse: jsDocParse,
  };

  function jsDocPreprocess(text: string, options: ParserOptions) {
    const tsPluginParser = findPluginByParser(parserName, options);

    if (!tsPluginParser) {
      return originalParser.preprocess
        ? originalParser.preprocess(text, options)
        : text;
    }

    const preprocess = tsPluginParser.preprocess ?? originalParser.preprocess;

    Object.assign(parser, {
      ...parser,
      ...tsPluginParser,
      preprocess: jsDocPreprocess,
      parse: jsDocParse,
    });

    return preprocess ? preprocess(text, options) : text;
  }

  return parser;
}
export type Options = Partial<JsdocOptions>;
