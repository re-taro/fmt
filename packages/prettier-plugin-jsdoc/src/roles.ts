import {
  ABSTRACT,
  ASYNC,
  AUGMENTS,
  AUTHOR,
  BORROWS,
  CALLBACK,
  CATEGORY,
  CLASS,
  CONSTANT,
  DEFAULT,
  DEFAULT_VALUE,
  DEPRECATED,
  DESCRIPTION,
  EXAMPLE,
  EXTENDS,
  EXTERNAL,
  FILE,
  FIRES,
  FLOW,
  FUNCTION,
  IGNORE,
  LICENSE,
  MEMBER,
  MEMBEROF,
  MODULE,
  NAMESPACE,
  OVERLOAD,
  OVERRIDE,
  PARAM,
  PRIVATE,
  PRIVATE_REMARKS,
  PROPERTY,
  PROVIDES_MODULE,
  REMARKS,
  RETURNS,
  SATISFIES,
  SEE,
  SINCE,
  TEMPLATE,
  THROWS,
  TODO,
  TYPE,
  TYPE_PARAM,
  TYPEDEF,
  VERSION,
  YIELDS,
} from "./tags";

const TAGS_SYNONYMS = {
  arg: PARAM,
  argument: PARAM,
  const: CONSTANT,
  constructor: CLASS,
  desc: DESCRIPTION,
  emits: FIRES,
  examples: EXAMPLE,
  exception: THROWS,
  fileoverview: FILE,
  func: FUNCTION,
  host: EXTERNAL,
  method: FUNCTION,
  overview: FILE,
  params: PARAM,
  prop: PROPERTY,
  return: RETURNS,
  var: MEMBER,
  virtual: ABSTRACT,
  yield: YIELDS,
  hidden: IGNORE,
};

const TAGS_DEFAULT = [DEFAULT, DEFAULT_VALUE];

const TAGS_NAMELESS = [
  BORROWS,
  CATEGORY,
  DEPRECATED,
  DESCRIPTION,
  EXAMPLE,
  EXTENDS,
  LICENSE,
  MODULE,
  NAMESPACE,
  OVERLOAD,
  OVERRIDE,
  PRIVATE_REMARKS,
  REMARKS,
  RETURNS,
  SINCE,
  THROWS,
  TODO,
  YIELDS,
  FILE,
  ...TAGS_DEFAULT,
];

const TAGS_TYPELESS = [
  BORROWS,
  BORROWS,
  DEPRECATED,
  DESCRIPTION,
  EXAMPLE,
  IGNORE,
  LICENSE,
  MODULE,
  NAMESPACE,
  OVERLOAD,
  OVERRIDE,
  PRIVATE_REMARKS,
  REMARKS,
  SINCE,
  TODO,
  FILE,
];

const TAGS_PEV_FORMATE_DESCRIPTION = [
  BORROWS,
  ...TAGS_DEFAULT,
  MEMBEROF,
  MODULE,
  SEE,
];

const TAGS_DESCRIPTION_NEEDED = [
  BORROWS,
  CATEGORY,
  DESCRIPTION,
  EXAMPLE,
  PRIVATE_REMARKS,
  REMARKS,
  SINCE,
  TODO,
];

const TAGS_TYPE_NEEDED = [
  EXTENDS,
  PARAM,
  PROPERTY,
  RETURNS,
  THROWS,
  TYPE,
  SATISFIES,
  TYPEDEF,
  YIELDS,
];

const TAGS_VERTICALLY_ALIGN_ABLE = [
  EXTENDS,
  PARAM,
  PROPERTY,
  RETURNS,
  THROWS,
  TYPE,
  SATISFIES,
  TYPEDEF,
  YIELDS,
];

const TAGS_GROUP_HEAD = [CALLBACK, TYPEDEF];
const TAGS_GROUP_CONDITION = [
  ...TAGS_GROUP_HEAD,
  TYPE,
  PROPERTY,
  PARAM,
  RETURNS,
  YIELDS,
  THROWS,
];

const TAGS_ORDER = [
  REMARKS,
  PRIVATE_REMARKS,
  PROVIDES_MODULE,
  MODULE,
  LICENSE,
  FLOW,
  ASYNC,
  PRIVATE,
  IGNORE,
  MEMBEROF,
  VERSION,
  FILE,
  AUTHOR,
  DEPRECATED,
  SINCE,
  CATEGORY,
  DESCRIPTION,
  EXAMPLE,
  ABSTRACT,
  AUGMENTS,
  CONSTANT,
  ...TAGS_DEFAULT,
  EXTERNAL,
  OVERLOAD,
  FIRES,
  TEMPLATE,
  TYPE_PARAM,
  FUNCTION,
  NAMESPACE,
  BORROWS,
  CLASS,
  EXTENDS,
  MEMBER,
  TYPEDEF,
  TYPE,
  SATISFIES,
  PROPERTY,
  CALLBACK,
  PARAM,
  YIELDS,
  RETURNS,
  THROWS,
  "other",
  SEE,
  TODO,
];

export {
  TAGS_DEFAULT,
  TAGS_DESCRIPTION_NEEDED,
  TAGS_GROUP_CONDITION,
  TAGS_GROUP_HEAD,
  TAGS_NAMELESS,
  TAGS_ORDER,
  TAGS_PEV_FORMATE_DESCRIPTION,
  TAGS_SYNONYMS,
  TAGS_TYPE_NEEDED,
  TAGS_TYPELESS,
  TAGS_VERTICALLY_ALIGN_ABLE,
};
