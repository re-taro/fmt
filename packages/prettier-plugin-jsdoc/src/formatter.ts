import type { Image, Link, List, Root, RootContent, Text } from "mdast";
import { fromMarkdown } from "mdast-util-from-markdown";
import type { BuiltInParserName } from "prettier";
import { format } from "prettier";

import { TAGS_PEV_FORMATE_DESCRIPTION } from "./roles";
import { DESCRIPTION, EXAMPLE, TODO } from "./tags";
import type { AllOptions } from "./types";
import { capitalizer, formatCode } from "./utils";

const TABLE = "2@^5!~#sdE!_TABLE";

interface DescriptionEndLineParams {
  tag: string;
  isEndTag: boolean;
}

function parserSynonyms(lang: string): BuiltInParserName[] {
  switch (lang) {
    case "js":
    case "javascript":
    case "jsx": {
      return ["babel", "babel-flow", "vue"];
    }
    case "ts":
    case "typescript":
    case "tsx": {
      return ["typescript", "babel-ts", "angular"];
    }
    case "json":
    case "css": {
      return ["css"];
    }
    case "less": {
      return ["less"];
    }
    case "scss": {
      return ["scss"];
    }
    case "html": {
      return ["html"];
    }
    case "yaml": {
      return ["yaml"];
    }
    default: {
      return ["babel"];
    }
  }
}

function descriptionEndLine({
  tag,
  isEndTag,
}: DescriptionEndLineParams): string {
  if ([DESCRIPTION, EXAMPLE, TODO].includes(tag) && !isEndTag) {
    return "\n";
  }

  return "";
}

interface FormatOptions {
  tagStringLength?: number;
  beginningSpace: string;
}

/**
 * Trim, make single line with capitalized text. Insert dot if flag for it is
 * set to true and last character is a word character
 *
 * @private
 */
async function formatDescription(
  tag: string,
  text: string,
  options: AllOptions,
  formatOptions: FormatOptions,
): Promise<string> {
  if (!text) return text;

  const { printWidth } = options;
  const { tagStringLength = 0, beginningSpace } = formatOptions;

  text = text.replace(/^(\d+)[-][\s|]+/g, "$1. "); // Start
  text = text.replace(/\n+(\s*\d+)[-][\s]+/g, "\n$1. ");

  const fencedCodeBlocks = text.matchAll(/```\S*?\n[\s\S]+?```/gm);
  const indentedCodeBlocks = text.matchAll(
    /^\r?\n^(?:(?:(?:[ ]{4}|\t).*(?:\r?\n|$))+)/gm,
  );
  const allCodeBlocks = [...fencedCodeBlocks, ...indentedCodeBlocks];
  const tables: string[] = [];
  text = text.replace(
    /((\n|^)\|[\s\S]*?)((\n[^|])|$)/g,
    (code, _1, _2, _3, _, offs: number) => {
      for (const block of allCodeBlocks) {
        if (
          block.index !== undefined &&
          block.index <= offs + 1 &&
          offs + code.length + 1 <= block.index + block[0].length
        ) {
          return code;
        }
      }

      code = _3 ? code.slice(0, -1) : code;

      tables.push(code);

      return `\n\n${TABLE}\n\n${_3 ? _3.slice(1) : ""}`;
    },
  );

  if (
    options.jsdocCapitalizeDescription &&
    !TAGS_PEV_FORMATE_DESCRIPTION.includes(tag)
  ) {
    text = capitalizer(text);
  }

  text = `${tagStringLength ? `${"!".repeat(tagStringLength - 1)}?` : ""}${
    text.startsWith("```") ? "\n" : ""
  }${text}`;

  let tableIndex = 0;

  text = text.replace(
    new RegExp("\\n" + `[\u0020]{${beginningSpace.length}}`, "g"),
    "\n",
  );

  const rootAst = fromMarkdown(text);

  async function stringifyASTWithoutChildren(
    mdAst: RootContent | Root,
    intention: string,
    parent: RootContent | Root | null,
  ) {
    if (mdAst.type === "inlineCode") {
      return `\`${mdAst.value}\``;
    }

    if (mdAst.type === "code") {
      let result = mdAst.value || "";
      let _intention = intention;

      if (result) {
        if (mdAst.lang) {
          const supportParsers = parserSynonyms(mdAst.lang.toLowerCase());
          const parser = supportParsers?.includes(options.parser as any)
            ? options.parser
            : supportParsers?.[0] || mdAst.lang;

          result = await formatCode(result, intention, {
            ...options,
            parser,
            jsdocKeepUnParseAbleExampleIndent: true,
          });
        } else if (options.jsdocPreferCodeFences || false) {
          result = await formatCode(result, _intention, {
            ...options,
            jsdocKeepUnParseAbleExampleIndent: true,
          });
        } else {
          _intention = intention + " ".repeat(4);

          result = await formatCode(result, _intention, {
            ...options,
            jsdocKeepUnParseAbleExampleIndent: true,
          });
        }
      }
      const addFence = options.jsdocPreferCodeFences || !!mdAst.lang;
      result = addFence ? result : result.trimEnd();

      return result
        ? addFence
          ? `\n\n${_intention}\`\`\`${mdAst.lang ?? ""}${result}\`\`\``
          : `\n${result}`
        : "";
    }

    if ((mdAst as Text).value === TABLE) {
      if (parent) {
        (parent as any).costumeType = TABLE;
      }

      if (tables.length > 0) {
        let result = tables?.[tableIndex] || "";
        tableIndex++;
        if (result) {
          result = (
            await format(result, {
              ...options,
              parser: "markdown",
            })
          ).trim();
        }

        return `${
          result
            ? `\n\n${intention}${result.split("\n").join(`\n${intention}`)}`
            : (mdAst as Text).value
        }`;
      }
    }

    if (mdAst.type === "break") {
      return "\\\n";
    }

    return (
      (mdAst as Text).value ??
      (mdAst as Link).title ??
      (mdAst as Image).alt ??
      ""
    );
  }

  async function stringyfy(
    mdAst: RootContent | Root,
    intention: string,
    parent: RootContent | Root | null,
  ): Promise<string> {
    if (!Array.isArray((mdAst as Root).children)) {
      return stringifyASTWithoutChildren(mdAst, intention, parent);
    }

    return (
      await Promise.all(
        (mdAst as Root).children.map(async (ast, index) => {
          switch (ast.type) {
            case "listItem": {
              let _listCount = `\n${intention}- `;
              if (typeof (mdAst as List).start === "number") {
                const count = index + ((mdAst as List).start ?? 1);
                _listCount = `\n${intention}${count}. `;
              }

              const _intention = intention + " ".repeat(_listCount.length - 1);

              const result = (await stringyfy(ast, _intention, mdAst)).trim();

              return `${_listCount}${result}`;
            }

            case "list": {
              let end = "";
              if (
                tag !== DESCRIPTION &&
                mdAst.type === "root" &&
                index === mdAst.children.length - 1
              ) {
                end = "\n";
              }

              return `\n${await stringyfy(ast, intention, mdAst)}${end}`;
            }

            case "paragraph": {
              const paragraph = await stringyfy(ast, intention, parent);
              if ((ast as any).costumeType === TABLE) {
                return paragraph;
              }

              return `\n\n${paragraph
                .split("\\\n")
                .map((_paragraph) => {
                  const links: string[] = [];
                  _paragraph = _paragraph.replace(
                    /{@(link|linkcode|linkplain)[\s]((?:[^{}])*)}/g,
                    (_, tag: string, link: string) => {
                      links.push(link);

                      return `{@${tag}${"_".repeat(link.length)}}`;
                    },
                  );

                  _paragraph = _paragraph.replace(/\s+/g, " ");

                  if (
                    options.jsdocCapitalizeDescription &&
                    !TAGS_PEV_FORMATE_DESCRIPTION.includes(tag)
                  )
                    _paragraph = capitalizer(_paragraph);
                  if (options.jsdocDescriptionWithDot)
                    _paragraph = _paragraph.replace(/([\w\p{L}])$/u, "$1.");

                  let result = breakDescriptionToLines(
                    _paragraph,
                    printWidth,
                    intention,
                  );

                  result = result.replace(
                    /{@(link|linkcode|linkplain)([_]+)}/g,
                    (original: string, tag: string, underline: string) => {
                      const link = links[0];

                      if (link.length === underline.length) {
                        links.shift();

                        return `{@${tag} ${link}}`;
                      }

                      return original;
                    },
                  );

                  return result;
                })
                .join("\\\n")}`;
            }

            case "strong": {
              return `**${await stringyfy(ast, intention, mdAst)}**`;
            }

            case "emphasis": {
              return `_${await stringyfy(ast, intention, mdAst)}_`;
            }

            case "heading": {
              return `\n\n${intention}${"#".repeat(
                ast.depth,
              )} ${await stringyfy(ast, intention, mdAst)}`;
            }

            case "link":
            case "image": {
              return `[${await stringyfy(ast, intention, mdAst)}](${ast.url})`;
            }

            case "linkReference": {
              return `[${await stringyfy(ast, intention, mdAst)}][${
                ast.label
              }]`;
            }
            case "definition": {
              return `\n\n[${ast.label}]: ${ast.url}`;
            }

            case "blockquote": {
              const paragraph = await stringyfy(ast, "", mdAst);

              return `${intention}> ${paragraph
                .trim()
                .replace(/(\n+)/g, `$1${intention}> `)}`;
            }
          }

          return stringyfy(ast, intention, mdAst);
        }),
      )
    ).join("");
  }

  let result = await stringyfy(rootAst, beginningSpace, null);

  result = result.replace(/^[\s\n]+/g, "");
  result = result.replace(/^(?:[!]+\?)/g, "");

  return result;
}

function breakDescriptionToLines(
  desContent: string,
  maxWidth: number,
  beginningSpace: string,
): string {
  let str = desContent.trim();

  if (!str) {
    return str;
  }

  let result = "";
  while (str.length > maxWidth) {
    let sliceIndex = str.lastIndexOf(
      " ",
      str.startsWith("\n") ? maxWidth + 1 : maxWidth,
    );
    if (sliceIndex <= beginningSpace.length)
      sliceIndex = str.indexOf(" ", beginningSpace.length + 1);

    if (sliceIndex === -1) sliceIndex = str.length;

    result += str.slice(0, Math.max(0, sliceIndex));
    str = str.slice(Math.max(0, sliceIndex + 1));
    if (str) {
      str = `${beginningSpace}${str}`;
      str = `\n${str}`;
    }
  }

  result += str;

  return `${beginningSpace}${result}`;
}

export type { FormatOptions };
export { descriptionEndLine, formatDescription };
