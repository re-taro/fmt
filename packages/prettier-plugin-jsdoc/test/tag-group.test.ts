import * as prettier from "prettier";
import { AllOptions } from "../src/types";
import { test, expect } from "vitest";

function subject(code: string, options: Partial<AllOptions> = {}) {
  return prettier.format(code, {
    parser: "babel",
    plugins: ["@re-taro/prettier-plugin-jsdoc"],
    ...options,
  } as AllOptions);
}

test("Tag group", async () => {
  const result = await subject(
    `
  /**
   * Aliquip ex proident tempor eiusmod aliquip amet. Labore commodo nulla tempor
   * consequat exercitation incididunt non. Duis laboris reprehenderit proident
   * proident.
   * @see {@link http://acme.com}
   * @example
   *   const foo = 0;
   *   const a = "";
   *   const b = "";
   *
   * @param id A test id.
   * @throws Minim sit ad commodo ut dolore magna magna minim consequat. Ex
   *   consequat esse incididunt qui voluptate id voluptate quis ex et. Ullamco
   *   cillum nisi amet fugiat.
   * @return Minim sit a.
   */

`,
    {
      jsdocSeparateTagGroups: true,
    },
  );

  expect(result).toMatchSnapshot();
});

test("space after unknownTag", async () => {
  function _subject(str: string) {
    return subject(str, {
      arrowParens: "always",
      bracketSameLine: false,
      bracketSpacing: true,
      embeddedLanguageFormatting: "auto",
      endOfLine: "lf",
      htmlWhitespaceSensitivity: "css",
      insertPragma: false,
      jsxSingleQuote: true,
      printWidth: 180,
      proseWrap: "preserve",
      quoteProps: "preserve",
      requirePragma: false,
      semi: true,
      singleAttributePerLine: false,
      singleQuote: true,
      tabWidth: 4,
      trailingComma: "all",
      useTabs: true,
      vueIndentScriptAndStyle: true,

      jsdocAddDefaultToDescription: false,
      jsdocCapitalizeDescription: true,
      jsdocDescriptionTag: false,
      jsdocDescriptionWithDot: true,
      jsdocKeepUnParseAbleExampleIndent: false,
      jsdocLineWrappingStyle: "greedy",
      jsdocPreferCodeFences: false,
      jsdocPrintWidth: 120,
      jsdocSeparateReturnsFromParam: false,
      jsdocSeparateTagGroups: true,
      jsdocCommentLineStrategy: "multiline",
      jsdocSpaces: 1,
      jsdocVerticalAlignment: true,
      tsdoc: false,
    });
  }
  const result = await _subject(`
   /**
    *
    * A description.
    *
    * @unknownTag A note.
    *
    * @see http://acme.com
    */
   `);

  expect(await _subject(await _subject(result))).toMatchSnapshot();
});

test("Inconsistant formatting", async () => {
  const result = await subject(
    `
    /**
     * Aliquip ex proident tempor eiusmod aliquip amet. Labore commodo nulla tempor
     * consequat exercitation incididunt non. Duis laboris reprehenderit proident
     * proident.
     *
     * @example
     *   const foo = 0;
     *
     *
     * @param id A test id.
     *
     * @throws Minim sit ad commodo ut dolore magna magna minim consequat. Ex
     *   consequat esse incididunt qui voluptate id voluptate quis ex et. Ullamco
     *   cillum nisi amet fugiat.
     * @see {@link http://acme.com}
     */
`,
    {
      jsdocSeparateTagGroups: true,
    },
  );

  expect(result).toMatchSnapshot();
});
