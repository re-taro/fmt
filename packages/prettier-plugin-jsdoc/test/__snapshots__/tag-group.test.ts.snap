// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Inconsistant formatting 1`] = `
"/**
 * Aliquip ex proident tempor eiusmod aliquip amet. Labore commodo nulla
 * tempor consequat exercitation incididunt non. Duis laboris reprehenderit
 * proident proident.
 *
 * @example
 *   const foo = 0;
 *
 * @param id A test id.
 *
 * @throws Minim sit ad commodo ut dolore magna magna minim consequat. Ex
 *   consequat esse incididunt qui voluptate id voluptate quis ex et.
 *   Ullamco cillum nisi amet fugiat.
 *
 * @see {@link http://acme.com}
 */
"
`;

exports[`Tag group 1`] = `
"/**
 * Aliquip ex proident tempor eiusmod aliquip amet. Labore commodo nulla
 * tempor consequat exercitation incididunt non. Duis laboris reprehenderit
 * proident proident.
 *
 * @example
 *   const foo = 0;
 *   const a = \\"\\";
 *   const b = \\"\\";
 *
 * @param id A test id.
 *
 * @returns Minim sit a.
 *
 * @throws Minim sit ad commodo ut dolore magna magna minim consequat. Ex
 *   consequat esse incididunt qui voluptate id voluptate quis ex et. Ullamco
 *   cillum nisi amet fugiat.
 *
 * @see {@link http://acme.com}
 */
"
`;

exports[`space after unknownTag 1`] = `
"/**
 * A description.
 *
 * @unknownTag A note.
 *
 * @see http://acme.com
 */
"
`;
